### PR TITLE
Fix double sharding in ShuffleBuffer

### LINKED ISSUE #7986: ShuffleBuffer not returning all patches
**Describe the bug**
While following the tutorial https://github.com/Project-MONAI/tutorials/blob/main/modules/2d_slices_from_3d_training.ipynb and implementing parts of it to my project especially when it comes to transforming Dataset containing 3D to 2D patches I encountered an issue with ShuffleBuffer.

`def create_dataset_2D_ds(ds, keys: list, trans2d: list) -> monai.data.ShuffleBuffer:
    # ds = CacheDataset(data=data_dicts, transform=transforms)
    patch_func = monai.data.PatchIterd(
        keys=keys, patch_size=(None, None, 1), start_pos=(0, 0, 0)
    )

    patch_transform = Compose([
            SqueezeDimd(keys=keys, dim=-1)
            ] + trans2d)

    patch_ds = monai.data.GridPatchDataset( # -> GridPatchDataset contains 531 entries
        data=ds, patch_iter=patch_func, transform=patch_transform, with_coordinates=False
    )
    sb = monai.data.ShuffleBuffer(patch_ds, buffer_size=200, seed=42, epochs=1) # -> ShuffleBuffer returning only 133 of them
    return patch_ds`

The problem is described above in the comments, I know that the GridPatchDataset contains 531 entries, but using Shuffle Buffer in DataLoader will result in 133.

**Expected behavior**
Being able to Iterate with ShuffleBuffer through entire GridPatchDataset

### PR DESCRIPTION
**Pull request by yinkev**
Fixes #7986.

### Description

`ShuffleBuffer` now distinguishes sources that already partition their stream using `torch.utils.data.get_worker_info`. MONAI `IterableDataset` sources are recognized automatically, while an explicit `source_shards_by_worker` override supports generic PyTorch or unusual custom iterable sources.

The default auto-detects MONAI iterable sources, so the `GridPatchDataset` pipeline reported in #7986 works without a caller change. `True` prevents a second partition for any worker-aware source; `False` preserves the outer worker partition for an explicitly unsharded iterable source.

This avoids using class identity as a proxy for sharding behavior: MONAI subclasses can be unsharded, while generic PyTorch iterable sources can shard themselves.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### Validation

- the original double-sharding reproduction loses samples when an already-sharded source is partitioned again;
- tests cover both explicitly worker-sharded MONAI/PyTorch sources and an unsharded MONAI subclass whose default behavior must remain unchanged;
- affected data tests: 7 passed;
- real two-worker fork-context `DataLoader` contract: the default MONAI-source path returns all 40 inputs exactly once;
- Black, isort, Ruff, DCO, and `git diff --check` pass.

### CODE REVIEW

**Comment by yinkev** [comment]
@coderabbitai review

File: tests/data/test_shuffle_buffer.py [hunk 1]
```diff
@@ -37,6 +61,48 @@ def test_shape(self):
...
+        outputs = []
+        for worker_id in range(2):
+            source = _UnshardedMonaiIterable(range(40))
+            buffer = ShuffleBuffer(source, buffer_size=8, seed=7, source_shards_by_worker=False)
+            worker_info = SimpleNamespace(num_workers=2, id=worker_id)
+            with patch("monai.data.iterable_dataset.get_worker_info", return_value=worker_info):
+                outputs.extend(buffer)
+
+        self.assertEqual(len(outputs), 40)
+        self.assertEqual(set(outputs), set(range(40)))
```
[the remark this answers is not in the document]
Comment (yinkev) [reply tests/data/test_shuffle_buffer.py]: Renamed the test and revised its docstring to describe the explicitly configured `source_shards_by_worker=False` path. The setup and assertions are unchanged; the affected data tests and format checks pass on the updated head.

File: monai/data/iterable_dataset.py [hunk 2]
```diff
@@ -97,11 +102,22 @@ def run():
...
+        epochs: int = 1,
+        source_shards_by_worker: bool | None = None,
+    ) -> None:
         super().__init__(data=data, transform=transform)
         self.size = buffer_size
         self.seed = seed
         self.epochs = epochs
+        self.source_shards_by_worker = (
+            isinstance(data, IterableDataset) if source_shards_by_worker is None else source_shards_by_worker
+        )
```
[the remark this answers is not in the document]
Comment (yinkev) [reply monai/data/iterable_dataset.py:156]: Added a direct Google-style constructor docstring documenting every argument, including `source_shards_by_worker`, while leaving initialization behavior unchanged. I omitted a `Returns` section because `__init__` returns `None` by language contract and the adjacent MONAI `IterableDataset.__init__` follows the same Args-only convention. The affected data tests and formatting checks pass on the new head.

File: monai/data/iterable_dataset.py [hunk 3]
```diff
@@ -197,6 +247,8 @@ class CSVIterableDataset(IterableDataset):
 
     """
 
+    _shards_by_worker = True
+
```
[the remark this answers is not in the document]
Comment (yinkev) [reply monai/data/iterable_dataset.py:263]: Fixed in 3a4e9759. The shuffled branch now exclusively yields from ShuffleBuffer, while the non-shuffled path remains in the else branch. A one-shot DataFrame-chunk regression previously raised on the second source iteration; it now asserts one source iteration and both unique records. The CSV + ShuffleBuffer tests pass (7 passed), the broader CSV/shuffle/grid/iterable slice exits cleanly, and Ruff check/format plus git diff --check pass.
